File: c/3105_longest_strictly_increasing_or_strictly_decreasing_subarray.c
int max3(int x, int y, int z) {
    if ((x > y) && (x > z)) {
        return x;
    } else if ((y > x) && (y > z)) {
        return y;
    } else {
        return z;
    }
}

int longestMonotonicSubarray(int* nums, int numsSize) {
    int longest_increasing;
    int longest_decreasing;
    int max_length;

    longest_increasing = 1;
    longest_decreasing = 1;
    max_length = 1;

    for (int i = 0; i < numsSize - 1; i++) {
        if (nums[i] > nums[i+1]) {
            longest_decreasing += 1;
            longest_increasing = 1;
        } else if (nums[i] < nums[i+1]) {
            longest_increasing += 1;
            longest_decreasing = 1;
        } else {
            longest_increasing = 1;
            longest_decreasing = 1;
        }
        max_length = max3(longest_increasing, longest_decreasing, max_length);
    }

    return max_length;
}

File: python/3105_longest_strictly_increasing_or_strictly_decreasing_subarray.py
class Solution:
    def longestMonotonicSubarray(self, nums: List[int]) -> int:
        n = len(nums)
        decreasing_subsequences = [
            list(yield_while_decreasing(nums, i))
            for i in range(n)
        ]
        increasing_subsequences = [
            list(yield_while_increasing(nums, i))
            for i in range(n)
        ]
        longest_decreasing = max(len(s) for s in decreasing_subsequences)
        longest_increasing = max(len(s) for s in increasing_subsequences)
        return max(longest_decreasing, longest_increasing)


def yield_while_decreasing(xs, start):
    yield from yield_while(xs, start, operator.lt)


def yield_while_increasing(xs, start):
    yield from yield_while(xs, start, operator.gt)


def yield_while(xs, start, binop):
    current = xs[start]
    yield current
    for i in range(start+1, len(xs)):
        if binop(xs[i], current):
            current = xs[i]
            yield current
        else:
            return
